### PR TITLE
Electron-256 (Changed the implementation from modal to electron window)

### PR DESCRIPTION
**Pull request by KiranNiranjan**
## Description
Previous implementation (`MediaSelectorModal`) had issues with the screen picker not showing when the meeting is in fullscreen mode. Changed the implementation from modal to electron browser window 

#### This PR fixes the following Tickets:
[ELECTRON-256](https://perzoinc.atlassian.net/browse/ELECTRON-256)
[ELECTRON-197](https://perzoinc.atlassian.net/browse/ELECTRON-197)

## Approach
How does this change address the problem?
- #### Problem with the code: Clicking on the `share screen` button in the meeting does not show the screen picker modal when the meeting is in fullscreen mode.
- #### Fix:  Changing the implementation from modal to electron browser window fixes the issue

#### Before
screen shot 2018-02-02 at 10 08 32 am

#### After
screen shot 2018-02-21 at 2 39 21 pm

## Learning
N/A

#### Blog Posts
N/A

## Open Questions if any and Todos
- [ ] Unit-Tests (Not Applicable)
- [ ] Documentation (No Changes)
- [ ] Automation-Tests
When solved, check the box and explain the answer.

@VikasShashidhar & @VishwasShashidhar Please Review. Thanks 😄

### CODE REVIEW

**Comment by VishwasShashidhar** [comment]
I think we should remove link to ELECTRON-310 because the fix was applied to R50 anyway.

File: js/desktopCapturer/getSources.js [hunk 1]
```diff
@@ -68,22 +72,24 @@ function getSources(options, callback) {
...
+        ipcRenderer.send(apiName, {
+            cmd: apiCmds.openScreenPickerWindow,
+            sources: sources,
+            id: id
+        });
+
+        function successCallback(e, source) {
+            // Cleaning up the event listener to prevent memory leaks
+            if (!source) {
+                ipcRenderer.removeListener('stat-share' + id, func);
```
Comment (VishwasShashidhar) [inline js/desktopCapturer/getSources.js]: `stat-share`, is that a typo?

File: js/desktopCapturer/index.js [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+const log = require('../log.js');
+const logLevels = require('../enums/logLevels.js');
+const { isMac, isWindowsOS } = require('./../utils/misc.js');
+
+let screenPickerWindow;
+let preloadWindow;
+let eventId;
+
+let windowConfig = {
+    width: 580,
```
Comment (VishwasShashidhar) [inline js/desktopCapturer/index.js]: can we get this dynamically?

File: js/desktopCapturer/index.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    height: isMac ? 519 : 523,
```
Comment (VishwasShashidhar) [inline js/desktopCapturer/index.js]: can we get these dynamically?

File: js/desktopCapturer/index.js [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    show: false,
+    modal: true,
+    frame: false,
+    autoHideMenuBar: true,
+    resizable: false,
+    alwaysOnTop: true,
+    webPreferences: {
+        preload: path.join(__dirname, 'renderer.js'),
+        sandbox: true,
```
Comment (VishwasShashidhar) [inline js/desktopCapturer/index.js]: is this needed?

File: js/desktopCapturer/index.js [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    screenPickerWindow = new BrowserWindow(windowConfig);
+    screenPickerWindow.setVisibleOnAllWorkspaces(true);
+    screenPickerWindow.loadURL(getTemplatePath());
+
+    screenPickerWindow.once('ready-to-show', () => {
+        screenPickerWindow.show();
+    });
+
+    screenPickerWindow.webContents.on('did-finish-load', () => {
+        screenPickerWindow.webContents.send('desktop-capturer-sources', sources, isWindowsOS);
```
Comment (VishwasShashidhar) [inline js/desktopCapturer/index.js]: what is the reason for checking windows os?

File: js/desktopCapturer/renderer.js [hunk 6]
```diff
@@ -0,0 +1,264 @@
+'use strict';
+const { ipcRenderer } = require('electron');
+
+const screenRegExp = new RegExp(/^Screen \d+$/gmi);
+const PAGE_DOWN = 34,
```
Comment (VishwasShashidhar) [inline js/desktopCapturer/renderer.js]: can we add code comments for these?

File: js/desktopCapturer/getSources.js (hunk 1 above)
Comment (KiranNiranjan) [reply js/desktopCapturer/getSources.js]: Yes. I've fixed it 👍

File: js/desktopCapturer/index.js (hunk 2 above)
Comment (KiranNiranjan) [reply js/desktopCapturer/index.js]: Dynamically with reference to display resolution or HTML content?

File: js/desktopCapturer/index.js (hunk 5 above)
Comment (KiranNiranjan) [reply js/desktopCapturer/index.js]: `isWindowsOS`is required to add the window borders only when the app is running on Windows OS because creating a frameless window electron app loses its native borders and drop shadow.

File: js/desktopCapturer/index.js (hunk 4 above)
Comment (KiranNiranjan) [reply js/desktopCapturer/index.js]: Yes, It makes this `browserWindow` more secure 😄

File: js/desktopCapturer/renderer.js [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+    shareButton.innerText = 'Share'
+}
+
+/**
+ * Loops through the items and removes
+ * the selected class property
+ */
+function highlightSelectedSource() {
+    let items = document.getElementsByClassName('item-container');
+    for (let i = 0, len = items.length; i < len; i++) {
```
Comment (VikasShashidhar) [inline js/desktopCapturer/renderer.js]: Could be simplified to `for (let i = 0; i < items.length; i++)`

File: js/desktopCapturer/renderer.js [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+    if (shareButton && shareButton.classList[0] === 'share-button-disable') {
+        shareButton.innerText = text;
+    }
+}
+
+/**
+ * Method handles used key up event
+ * @param e
+ */
+function handleKeyUpPress(e) {
```
Comment (VikasShashidhar) [inline js/desktopCapturer/renderer.js]: Could be renamed to `handleKeyboardInteractions`

File: js/desktopCapturer/renderer.js [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+                updateSelectedSource(availableSources.length - 1);
+            }
+            break;
+        case ARROW_DOWN:
+            updateSelectedSource(2);
+            break;
+        case ENTER_KEY:
+            startShare();
+            break;
+        case ESCAPE_KEY:
```
Comment (VikasShashidhar) [inline js/desktopCapturer/renderer.js]: We could also consider implementing for `tab`

File: js/desktopCapturer/screen-picker.html [hunk 10]
```diff
@@ -0,0 +1,239 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <title>Screen Picker</title>
+    <style>
+
+        body {
+            margin: 0 auto;
+            font-family: "Helvetica Neue", sans-serif;
```
Comment (VikasShashidhar) [inline js/desktopCapturer/screen-picker.html]: Can we look at following the font fallback used in the client app?

File: js/desktopCapturer/screen-picker.html [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+            overflow: hidden;
+        }
+
+        .window-border {
+            border: 2px solid #444;
```
Comment (VikasShashidhar) [inline js/desktopCapturer/screen-picker.html]: Try to use the RGBA color coding. You'll get finer control

File: js/desktopCapturer/screen-picker.html [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+
+        span {
+            font-style: normal;
+            margin: 0 0 0 4px;
+            font-size: 1.2rem;
+            min-height: 13px;
+        }
+
+        input {
+            display: none;
```
Comment (VikasShashidhar) [inline js/desktopCapturer/screen-picker.html]: Why is this required?

File: js/desktopCapturer/screen-picker.html [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+            display: inline-block;
+            height: 150px;
+            margin: 5px;
+            position: relative;
+            width: 40%;
+            border: 1px solid rgba(0, 0, 0, .1);
+        }
+
+        .item-container:hover {
+            box-shadow: 0 1px 5px rgba(0, 0, 0, .1), 0 5px 5px rgba(0, 0, 0, .1);
```
Comment (VikasShashidhar) [inline js/desktopCapturer/screen-picker.html]: You could also make the cursor a pointer.

File: js/desktopCapturer/screen-picker.html [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+            color: rgba(0, 0, 0, 0.38);
+            background-color: rgba(255, 255, 255, 0);
+            text-transform: uppercase;
+            margin: 0;
+            padding: 8px 16px;
+        }
+
+        .cancel-button:hover {
+            background-color: rgba(255, 255, 255, 0);
+            color: black;
```
Comment (VikasShashidhar) [inline js/desktopCapturer/screen-picker.html]: Could use a cursor pointer. Better indicates to the user that it is actionable.

File: js/desktopCapturer/screen-picker.html [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+        #error-content span {
+            color: rgba(0, 0, 0, 0.38);
+        }
+
+    </style>
+</head>
+<body>
+<div class="content">
+    <div class="custom-window-title">
+        <span>Choose what you'd like to share</span>
```
Comment (VikasShashidhar) [inline js/desktopCapturer/screen-picker.html]: Try to use HTML unicodes for symbols - https://www.fileformat.info/info/unicode/char/27/index.htm

File: js/desktopCapturer/index.js (hunk 2 above)
Comment (KiranNiranjan) [reply js/desktopCapturer/index.js]: For creating a dynamic window with respect to HTML content. first, we need to open the window with some value for `width & height` and then we need to publish an event from `render process` to the `main process` when the DOM is loaded to resize the window.

Currently, the `width and height` of the window is exactly the same as the HTML content. I don't think that we need to dynamically update the window `width or height`

File: js/desktopCapturer/renderer.js (hunk 8 above)
Comment (KiranNiranjan) [reply js/desktopCapturer/renderer.js]: This method will be invoked only for `keyup` events

File: js/desktopCapturer/screen-picker.html (hunk 12 above)
Comment (KiranNiranjan) [reply js/desktopCapturer/screen-picker.html]: Input type for the tabs that I'm using is `radio` so in order to hide the radio button setting `display: none`

**Comment by KiranNiranjan** [comment]
@VikasShashidhar & @VishwasShashidhar Added a new API to support backward compatibility and deprecated the existing one. Please review. Thanks 😄

File: js/desktopCapturer/getSource.js [hunk 16]
```diff
@@ -0,0 +1,96 @@
+'use strict';
```
Comment (VikasShashidhar) [inline js/desktopCapturer/getSource.js]: Is this just a new API that does the same thing as the `getSources` API? If yes, why reinvent it?

Comment (KiranNiranjan) [reply js/desktopCapturer/getSource.js]: @VikasShashidhar `getSources` returns an array of sources whereas `getSource` returns single source selected by the user. We are keeping the `getSources` API to support older versions of Electron and Client app.

We can remove `getSources` after 2-3 sprints 😄
